### ISSUE TITLE
Worse performance in newer version when using relationships with dynamic length

### ISSUE THREAD

**Issue by ddaa2000** [neo4j/neo4j#12927]
I discovered that when a query becomes slower when using 4.4.10 compared to 4.3.17

**Neo4j Version:** 4.4.10
**Operating System:** Ubuntu 20.04
**API:** Java

# Steps to reproduce
1. Execute following queries to create the graph:[graph.txt](https://github.com/neo4j/neo4j/files/9418005/graph.txt)2. 
2. Then execute:
```
MATCH (n0 :L6)-[r0 *3..]-(n1 :L0), (n2)-[r1]->(n3) RETURN 0
```

4.4.10 takes more than a minute while 4.3.17 only takes less than a second.
Is this a performance bug? Or maybe this is because that neo4j has abondoned "..." in relationships ?

**Comment by HannesSandberg** [comment]
Thank you! We will look into this and report back to you.

**Comment by InverseFalcon** [comment]
Keep in mind this is a highly inefficient query, performing a cross product of every path matching the first match pattern with every single-relationship path in the graph `()-[]->()`. The planner has the ability to decide how to execute the query, and usually the more specific you can be with the pattern the more efficient it can be. Without anything to restrict the second part of that MATCH, it leaves it open for some operations that may not be efficient. 

Basically, there is no way this is any kind of useful valid use case, we would recommend using something that makes more sense.

Also, it would help to provide PROFILE plans for both versions for comparsion.

**Comment by pontusmelke** [comment]
@ddaa2000 I have run the query both on 4.4.10 and 4.3.17 and in neither version does the execution finish in less than a second? When running the query with EXPLAIN I get identical plans in 4.4 and in 4.3.

How did you conduct your test? Are you sure the data is identical on both databases? What does the plan look like when you run with `EXPLAIN` on the two databases?

As @InverseFalcon this is a very heavy query there are 198 nodes in your data matching `(n2)-[r1]->(n3)` and there are

 - 62 nodes that matches `(n0 :L6)-[r0*3]-(n1 :L0)`
 - 131 nodes that matches `(n0 :L6)-[r0*4]-(n1 :L0)`
 - 284 nodes that matches `(n0 :L6)-[r0*5]-(n1 :L0)`
 - 509 nodes that matches `(n0 :L6)-[r0*6]-(n1 :L0)`
 
and so on. And then you ask for the cartesian product between these results.

If you explain what you are trying to achieve with your query we can maybe help you optimize your query?

Best regards,
Pontus
